Perform proper cleanup for signals with 'Term' action

d-e-s-o opened the pull request:
Issue #3049 describes a case where terminating i3 by means of SIGTERM
causes it to leak the runtime directory and all its contents. There are
multiple issues at play: first, any cleanup handlers registered via
atexit are never invoked when a signal terminates the program (see
atexit(3)). Hence, the log SHM log cleanup performed in i3_exit is not
invoked in that case. Second, compared to the shutdown path for the
'exit' command, we do not unlink the UNIX domain socket we create,
causing it to be leaked as well. Third, a handler for SIGTERM is not
registered at all despite handle_signal claiming to be the handler for
all 'Term' signals.
This change addresses all three problems and results in a graceful exit
including cleanup to happen when we receive a signal with the default
action 'Term'. It addresses issue #3049.

stapelberg commented on src/main.c:
```diff
@@ -174,21 +174,34 @@ static void i3_exit(void) {
         fflush(stderr);
         shm_unlink(shmlogname);
     }
+    unlink(config.ipc_socket_path);
```
Looking at cmd_exit and cmd_restart in src/commands.c, I think we should unify that code (eventually?). For the time being, let’s at least also call ipc_shutdown and xcb_disconnect here

d-e-s-o replied:
Invoking `xcb_disconnect` here causes `t/229-cleanup-tmpdir.t` to fail. I don't have the time to investigate that right now, so I am not making this change. `ipc_shutdown` is not causing problems so I've added it.

stapelberg commented on src/main.c:
```diff
@@ -174,21 +174,34 @@ static void i3_exit(void) {
+static void handle_term_signal(int sig, siginfo_t *info, void *data) {
+    /* We exit gracefully here in the sense that cleanup handlers
+     * installed via atexit are invoked. */
+    exit(-sig);
```
I think `128+sig` is the exit code we should use, as per https://stackoverflow.com/a/1535733

d-e-s-o commented:
> The tests seem to hang, could you take a look at that please?
How can I access test-suite.log from the Travis run? Are we sure it's the newly added test that is hanging? I am asking because at 'next' top of tree complete-run.pl hang for me. I had to backtrack a bit to get the tests running (no changes applied). There is seemingly only a single test run (?) which suggests it being my test, but I cannot verify that.

stapelberg commented:
> How can I access test-suite.log from the Travis run? Are we sure it's the newly added test that is hanging?

The log is printed by default (see travis/run-tests.sh), but not when the run times out entirely.

You can use Docker to get the same environment which travis uses: https://hub.docker.com/u/i3wm/

> I am asking because at 'next' top of tree complete-run.pl hang for me. I had to backtrack a bit to get the tests running (no changes applied).

Can you report a separate issue for that? The tests work fine for me on the next branch. Are you sure you have all dependencies installed?

> There is seemingly only a single test run (?) which suggests it being my test, but I cannot verify that.

There is always only one test from the perspective of “make check”, and that test is to run i3’s test suite, which in turn runs all the individual files.

However, I checked out your branch on my computer, and while the tests work fine on `next`, they fail frequently with your changes.

d-e-s-o commented:
> The tests seem to hang, could you take a look at that please?

I believe that may be due to signal-unsafe code in the signal handler. I've reworked the change a bit to use libev's signal watchers which should not have that problem.

>> I am asking because at 'next' top of tree complete-run.pl hang for me. I had to backtrack a bit to get the tests running (no changes applied).

> Can you report a separate issue for that? The tests work fine for me on the next branch. Are you sure you have all dependencies installed?

I believe I may have been too impatient. The tests get stuck on occasion for 30s or so but if I wait long enough it will eventually succeed. Nevertheless on the CL I tested on this wasn't the case at all. Unfortunately, I do not recall which CL that was, as I rebased before creating a pull request. As that is only intermittent I won't dig into history...

> However, I checked out your branch on my computer, and while the tests work fine on next, they fail frequently with your changes.

I think the problem is with the refactoring happening in `testcases/lib/i3test.pm.in`. I reverted that. The reason I had it in the first place is because we are setting global state (`$id_pid`) from a normal function and in parallel runs this is causing problems because we kill a child twice (or so is my understanding). Now why it fails with the refactoring is still beyond me. There is something magical about `$id_pid`. The mere moving of its assignment is making tests fail (intermittently).

d-e-s-o commented:
Hi Michael, would you mind having a look at the latest version of the change? I'd like to get closure on this pull request, if possible.

stapelberg commented on src/main.c:
```diff
@@ -38,6 +38,9 @@ int listen_fds;
 static struct ev_prepare *xcb_prepare;
+static struct ev_signal signal_watchers[6];
+
+#define NUM_WATCHERS (sizeof(signal_watchers) / sizeof(signal_watchers[0]))
```
Given that we only ever create (but never modify) these watchers, I think we can do away with the variable and heap-allocate them.

d-e-s-o replied:
Are you asking me to introduce a memory leak into the program? ;-) My point is, all those allocations of `ev_io` objects in main.c are leaks and are being reported as such by leak/address sanitizer. Instead of introducing new leaks, I suggest we remove the existing ones to perhaps at some point have a run without leak reports.

If you have a problem with the global scope, however, I can make the variable a function local static. Let me know.

stapelberg commented on src/main.c:
```diff
@@ -174,21 +177,60 @@ static void i3_exit(void) {
+    for (size_t i = 0; i < NUM_WATCHERS; i++) {
+        ev_signal_start(main_loop, &signal_watchers[i]);
+        /* None of the signal handlers should hold a reference to the
+         * main loop. */
```
The comment documents the “what”, but not the “why”. Let’s make it a bit more descriptive. How about:

The signal handlers should not block ev_run from returning:

d-e-s-o replied:
Sure.

stapelberg commented on testcases/lib/i3test.pm.in:
```diff
@@ -770,6 +772,47 @@ sub exit_gracefully {
     undef $i3_pid;
 }
 
+=head2 exit_forcefully($pid, [ $signal ])
```
Since we only need this behavior in precisely one test case, could you move it into the test itself?

d-e-s-o replied:
The reason it is in i3test.pm.in is because of the whole business around `$i3_pid` (that I somewhat explained before). That is a variable that I don't think we should not expose outside of the module, Also, the `$tester` object we use seems to be local.

stapelberg commented:
Thanks!
